pootle/apps/virtualfolder/receivers.py

```diff
@@ -31,125 +30,74 @@ def update_vfolder_tree(vf, store):
 
 
-@receiver(pre_save, sender=VirtualFolder)
-def vfolder_unit_priority_presave_handler(**kwargs):
+@receiver(post_save, sender=VirtualFolder)
+def vfolder_save_handler(sender, instance, created, **kwargs):
     """Remove Units from VirtualFolder when vfolder changes
-
     - Check the original VirtualFolder object's locations
     - Check the new VirtualFolders object's locations
     - Remove any units from the VirtualFolder that are only in original's
       locations and reset Unit priority
     - Update VirtualFolderTree for any Stores that may have been affected
     """
-    instance = kwargs["instance"]
-    if instance.id is None:
-        return
-
-    original = VirtualFolder.objects.get(pk=instance.pk)
-
-    original_locations = set()
-    new_locations = set()
-
-    for location in original.all_locations:
-        for filename in original.filter_rules.split(","):
-            original_locations.add("".join([location, filename]))
-
+    locations = set()
     for location in instance.all_locations:
         for filename in instance.filter_rules.split(","):
-            new_locations.add("".join([location, filename]))
-
-    removed_locations = original_locations - new_locations
-
-    stores_affected = set()
-    for location in removed_locations:
-        # reindex these units without this vfolder
-        removed_units = (
-            Unit.objects.filter(store__pootle_path__startswith=location,
-                                vfolders=original))
-        for unit in removed_units.iterator():
-            unit.vfolders.remove(original)
-            stores_affected.add(unit.store)
-
-    for store in stores_affected:
-        update_vfolder_tree(original, store)
+            locations.add("".join([location, filename]))
+
+    stores_we_want = Store.objects.none()
+    for location in locations:
+        stores_we_want = stores_we_want | Store.objects.filter(
+            pootle_path__startswith=location)
+    to_remove = list(
+        instance.stores.exclude(
+            pk__in=stores_we_want.values_list("pk", flat=True)))
+    to_add = list(
+        stores_we_want.exclude(
+            pk__in=instance.stores.values_list("pk", flat=True)))
+    instance.stores.remove(*to_remove)
+    instance.stores.add(*to_add)
+    for store in to_add:
+        update_vfolder_tree(instance, store)
+        store.set_priority()
+    for store in to_remove:
         store.set_priority()
 
 
 @receiver(vfolder_post_save, sender=VirtualFolder)
-def vfolder_unit_priority_handler(**kwargs):
+def vfolder_store_priority_handler(**kwargs):
     """Set Unit priorities for VirtualFolder members on change
     """
     instance = kwargs["instance"]
-    stores = Store.objects.filter(
-        id__in=instance.units.values_list("store_id").distinct())
-    for store in stores:
+    for store in instance.stores.all():
         store.set_priority()
 
 
-@receiver(pre_save, sender=Unit)
-def vfolder_unit_resurrected(**kwargs):
```

Comment:
We probably will want to add some test to ensure a vfolder stats are changed when a unit is resurrected or obsoleted.